packages/server/modules/workspaces/services/management.ts

```diff
@@ -0,0 +1,187 @@
+} from '@/modules/workspaces/domain/operations'
+import { WorkspaceAdminRequiredError } from '@/modules/workspaces/errors/workspace'
+import { isUserLastWorkspaceAdmin } from '@/modules/workspaces/utils/roles'
+import { mapWorkspaceRoleToProjectRole } from '@/modules/workspaces/domain/roles'
+import { queryAllWorkspaceProjectsFactory } from '@/modules/workspaces/services/projects'
+
+type WorkspaceCreateArgs = {
+  workspaceInput: { name: string; description: string | null; logo: string | null }
+  userId: string
+}
+
+export const createWorkspaceFactory =
+  ({
+    upsertWorkspace,
+    upsertWorkspaceRole,
+    emitWorkspaceEvent,
+    storeBlob
+  }: {
+    upsertWorkspace: UpsertWorkspace
+    upsertWorkspaceRole: UpsertWorkspaceRole
+    storeBlob: StoreBlob
+    emitWorkspaceEvent: EmitWorkspaceEvent
+  }) =>
+  async ({ userId, workspaceInput }: WorkspaceCreateArgs): Promise<Workspace> => {
+    let logoUrl: string | null = null
+    if (workspaceInput.logo) {
+      logoUrl = await storeBlob(workspaceInput.logo)
+    }
+
+    const workspace = {
+      ...workspaceInput,
+      id: cryptoRandomString({ length: 10 }),
+      createdAt: new Date(),
+      updatedAt: new Date(),
+      logoUrl
+    }
+    await upsertWorkspace({ workspace })
+    // assign the creator as workspace administrator
+    await upsertWorkspaceRole({
+      userId,
+      role: Roles.Workspace.Admin,
+      workspaceId: workspace.id
+    })
+
+    await emitWorkspaceEvent({
+      eventName: WorkspaceEvents.Created,
+      payload: { ...workspace, createdByUserId: userId }
+    })
+    // emit a workspace created event
+
+    return workspace
+  }
+
+type WorkspaceRoleDeleteArgs = {
+  userId: string
+  workspaceId: string
+}
+
+export const deleteWorkspaceRoleFactory =
+  ({
+    getWorkspaceRoles,
+    deleteWorkspaceRole,
+    emitWorkspaceEvent,
+    getStreams,
+    revokeStreamPermissions
+  }: {
+    getWorkspaceRoles: GetWorkspaceRoles
+    deleteWorkspaceRole: DeleteWorkspaceRole
+    emitWorkspaceEvent: EmitWorkspaceEvent
+    getStreams: typeof repoGetStreams
+    revokeStreamPermissions: typeof repoRevokeStreamPermissions
+  }) =>
+  async ({
+    userId,
+    workspaceId
+  }: WorkspaceRoleDeleteArgs): Promise<WorkspaceAcl | null> => {
+    // Protect against removing last admin
+    const workspaceRoles = await getWorkspaceRoles({ workspaceId })
+    if (isUserLastWorkspaceAdmin(workspaceRoles, userId)) {
+      throw new WorkspaceAdminRequiredError()
+    }
+
+    // Perform delete
+    const deletedRole = await deleteWorkspaceRole({ userId, workspaceId })
+    if (!deletedRole) {
+      return null
+    }
+
+    // Delete workspace project roles
+    const projectIds: string[] = []
+    for await (const project of queryAllWorkspaceProjectsFactory({ getStreams })(
+      workspaceId
+    )) {
+      projectIds.push(project.id)
+    }
+    await Promise.all(
+      projectIds.map((streamId) => revokeStreamPermissions({ streamId, userId }))
```

Comment:
I mean this works, but ideally we'd want the generator to return the projects in batches, instead of one by one, so that you can then also `revokeStreamPermissions` in batches. As opposed to now collecting potentially hundreds or thousands of IDs and then invoking hundreds/thousands of concurrent promises that may kill the server.

```ts
for await (const projectsPage of projectPageGenerator) {
  await Promise.all(projectsPage.map(p => revokeStreamPermissions({ streamId: p.id, userId })))
}
```

That said, this is already better than what was here before

Reply:
I went back and forth but this is a very good reason to go by page, done